File: solutions/ruby/allergies/allergies.rb
class Allergies
  ALLERGENS = {
    "eggs" => 1,
    "peanuts" => 2,
    "shellfish" => 4,
    "strawberries" => 8,
    "tomatoes" => 16,
    "chocolate" => 32,
    "pollen" => 64,
    "cats" => 128,
  }.freeze

  def initialize(score)
    @score = score
  end

  def allergic_to?(allergen)
    @score & ALLERGENS[allergen] > 0
  end

  def list
    ALLERGENS.keys.filter { |allergen| allergic_to?(allergen) }
  end
end

File: solutions/javascript/allergies/allergies.js
/**
 * Allergies
 *
 * Given a person's allergy score, determine whether or not they're allergic
 * to a given item, and their full list of allergies.
 */
const ALLERGENS = [
  "eggs",
  "peanuts",
  "shellfish",
  "strawberries",
  "tomatoes",
  "chocolate",
  "pollen",
  "cats",
];

export class Allergies {
  constructor(score) {
    this.score = score;
  }

  /**
   * Give the whole list of allergies.
   * @returns {string[]} List of allergies.
   */
  list() {
    return ALLERGENS.filter((_, i) => this.score & (1 << i));
  }

  /**
   * Determine whether the user is allergic or not to some element.
   * @param {string} allergen Allergy type.
   * @returns {boolean} True if allergic to the given type.
   */
  allergicTo(allergen) {
    return this.list().includes(allergen);
  }
}
